ui/collapsible-panel.reel/collapsible-panel.js

```diff
@@ -0,0 +1,126 @@
+     * header to collapse or expand a panel by setting this property to true.
+     *
+     * @private
+     */
+    _isTransition: {
+        value: false
+    },
+
+    _isExpanded: {
+        value: null
+    },
+
+    /**
+     * CollapsiblePanel's data.
+     *
+     * @type {Object}
+     */
+    data: {
+        value: null
+    },
+
+    /**
+     * A ChoiceController that control expanded panel and multiple mode of CollapsiblePanel.
+     *
+     * @type {Object}
+     */
+    expansionController: {
+        set: function (value) {
+            if (value instanceof ChoiceController) {
+                this._expansionController = value;
+                if (this.getBinding("isExpanded")) {
+                    this.cancelBinding("isExpanded");
+                }
+                this.defineBinding("isExpanded", {
+                    "<->": "_expansionController.choice.data.has(data)"
+                });
+            } else {
+                throw new Error("expansionController must be a ChoiceController.");
+            }
+        }
+    },
+
+    /**
+     * Is the collapsible panel expanded or collapsed.
+     *
+     * @type {boolean}
+     */
+    isExpanded: {
+        set: function (value) {
+            if (value != this._isExpanded) {
+                this._isExpanded = value;
+                this.needsDraw = true;
+            }
+        },
+        get: function () {
+            return this._isExpanded;
+        }
+    },
+
+    prepareForActivationEvents: {
+        value: function () {
+            var handler = this.handleTransitionEnd.bind(this);
+            var bodyWrapper = this.element.children[1];
+            this._pressComposer.addEventListener("press", this, false);
+            // Listen to CSS transition call back.
+            bodyWrapper.addEventListener('webkitTransitionEnd', handler);
+            bodyWrapper.addEventListener('transitionend', handler);
+        }
+    },
+
+    handleTransitionEnd: {
+        value: function () {
+            this._isTransition = false;
+        }
+    },
+
+    /**
+     * Do expand or collapse when click on the header.
+     *
+     * @method
+     * @private
+     */
+    handlePress: {
+        value: function (event) {
+            // Only trigger css animation when clicking on header, this may need change base on how SelectionController
+            // handle multipleSelection and singleSelection.
+            this._isTransition = true;
+            this.isExpanded = !this.isExpanded;
+        }
+    },
+
+    enterDocument: {
+        value: function (firstTime) {
+            var header;
+            if (firstTime) {
+                header = this.element.children[0];
+                this._pressComposer = new PressComposer();
+                // Register composer here and add event in prepareForActivationEvents.
+                this.addComposerForElement(this._pressComposer, header);
+            }
+        }
+    }
+
+});
```

Comment:
Ouch... the hasOwnProperty() is more ugly than was the instanceof it replaces. I don't understand why any of these checks are necessary... What situation are you trying to prevent? Are you trying to make sure your code will work with all possible passed in values for expansionController? That's not possible: In a dynamic language like JavaScript no amount of checking will "make sure isExpanded binding can work correctly". Are you trying to catch errors developers are likely to make? How likely are they to make the mistake you're checking for, and in what way is the error message you're generating better than the error message they would have gotten without the extra type checking code?

My advice is: Remove all type checking and let the errors that naturally would occur if incorrect values are set occur. This will make the code simpler and less brittle.